File: mini-kuber.py
#!/usr/bin/env python3
"""
Mini-Kuber Python Template
Replace YAML hell with intelligent Python code

Usage:
1. Copy this file to your project
2. Modify the examples at the bottom
3. Run: python your-deployment.py
4. Your Kubernetes resources are deployed

No framework to install, no complexity, just copy and adapt.
"""

import json
import subprocess
import sys
import tempfile
from pathlib import Path
from typing import Dict, List, Any, Optional, Union
from dataclasses import dataclass, field


@dataclass
class Resource:
    """Base Kubernetes resource"""
    kind: str
    name: str
    namespace: str = "default"
    labels: Dict[str, str] = field(default_factory=dict)
    annotations: Dict[str, str] = field(default_factory=dict)
    
    def __post_init__(self):
        if not self.labels:
            self.labels = {"app": self.name, "managed-by": "mini-kuber"}


class Deployment(Resource):
    """Kubernetes Deployment made simple"""
    
    def __init__(self, name: str, image: str, replicas: int = 1, port: Optional[int] = None,
                 env: Optional[Dict[str, str]] = None, namespace: str = "default",
                 cpu_request: str = "100m", memory_request: str = "128Mi",
                 cpu_limit: str = "500m", memory_limit: str = "512Mi",
                 auto_heal: bool = True, **kwargs):
        
        super().__init__(kind="Deployment", name=name, namespace=namespace)
        self.image = image
        self.replicas = replicas
        self.port = port
        self.env = env or {}
        self.cpu_request = cpu_request
        self.memory_request = memory_request
        self.cpu_limit = cpu_limit
        self.memory_limit = memory_limit
        
        if auto_heal:
            self.labels["auto-heal"] = "true"
    
    def to_yaml(self) -> Dict[str, Any]:
        """Convert to Kubernetes YAML structure"""
        container = {
            "name": self.name,
            "image": self.image,
            "resources": {
                "requests": {"cpu": self.cpu_request, "memory": self.memory_request},
                "limits": {"cpu": self.cpu_limit, "memory": self.memory_limit}
            }
        }
        
        if self.port:
            container["ports"] = [{"containerPort": self.port}]
        
        if self.env:
            container["env"] = [{"name": k, "value": v} for k, v in self.env.items()]
        
        return {
            "apiVersion": "apps/v1",
            "kind": "Deployment",
            "metadata": {
                "name": self.name,
                "namespace": self.namespace,
                "labels": self.labels,
                "annotations": self.annotations
            },
            "spec": {
                "replicas": self.replicas,
                "selector": {"matchLabels": {"app": self.name}},
                "template": {
                    "metadata": {"labels": {"app": self.name, **self.labels}},
                    "spec": {"containers": [container]}
                }
            }
        }


class Service(Resource):
    """Kubernetes Service made simple"""
    
    def __init__(self, name: str, target: Union[Deployment, str], port: int = 80,
                 target_port: Optional[int] = None, service_type: str = "ClusterIP",
                 namespace: str = "default", **kwargs):
        
        super().__init__(kind="Service", name=name, namespace=namespace)
        self.target = target.name if isinstance(target, Deployment) else target
        self.port = port
        self.target_port = target_port or (target.port if isinstance(target, Deployment) else port)
        self.service_type = service_type
    
    def to_yaml(self) -> Dict[str, Any]:
        """Convert to Kubernetes YAML structure"""
        return {
            "apiVersion": "v1",
            "kind": "Service",
            "metadata": {
                "name": self.name,
                "namespace": self.namespace,
                "labels": self.labels,
                "annotations": self.annotations
            },
            "spec": {
                "selector": {"app": self.target},
                "ports": [{
                    "port": self.port,
                    "targetPort": self.target_port,
                    "protocol": "TCP"
                }],
                "type": self.service_type
            }
        }


class StatefulSet(Resource):
    """Kubernetes StatefulSet for databases and stateful apps"""
    
    def __init__(self, name: str, image: str, replicas: int = 1, port: Optional[int] = None,
                 storage: str = "10Gi", storage_class: str = "standard",
                 env: Optional[Dict[str, str]] = None, namespace: str = "default", **kwargs):
        
        super().__init__(kind="StatefulSet", name=name, namespace=namespace)
        self.image = image
        self.replicas = replicas
        self.port = port
        self.storage = storage
        self.storage_class = storage_class
        self.env = env or {}
    
    def to_yaml(self) -> Dict[str, Any]:
        """Convert to Kubernetes YAML structure"""
        container = {
            "name": self.name,
            "image": self.image,
            "volumeMounts": [{
                "name": f"{self.name}-storage",
                "mountPath": "/var/lib/data"
            }]
        }
        
        if self.port:
            container["ports"] = [{"containerPort": self.port}]
        
        if self.env:
            container["env"] = [{"name": k, "value": v} for k, v in self.env.items()]
        
        return {
            "apiVersion": "apps/v1",
            "kind": "StatefulSet",
            "metadata": {
                "name": self.name,
                "namespace": self.namespace,
                "labels": self.labels
            },
            "spec": {
                "serviceName": self.name,
                "replicas": self.replicas,
                "selector": {"matchLabels": {"app": self.name}},
                "template": {
                    "metadata": {"labels": {"app": self.name}},
                    "spec": {"containers": [container]}
                },
                "volumeClaimTemplates": [{
                    "metadata": {"name": f"{self.name}-storage"},
                    "spec": {
                        "accessModes": ["ReadWriteOnce"],
                        "storageClassName": self.storage_class,
                        "resources": {"requests": {"storage": self.storage}}
                    }
                }]
            }
        }


class ConfigMap(Resource):
    """Kubernetes ConfigMap for configuration"""
    
    def __init__(self, name: str, data: Dict[str, str], namespace: str = "default", **kwargs):
        super().__init__(kind="ConfigMap", name=name, namespace=namespace)
        self.data = data
    
    def to_yaml(self) -> Dict[str, Any]:
        return {
            "apiVersion": "v1",
            "kind": "ConfigMap",
            "metadata": {
                "name": self.name,
                "namespace": self.namespace,
                "labels": self.labels
            },
            "data": self.data
        }


class Secret(Resource):
    """Kubernetes Secret for sensitive data"""
    
    def __init__(self, name: str, data: Dict[str, str], secret_type: str = "Opaque",
                 namespace: str = "default", **kwargs):
        super().__init__(kind="Secret", name=name, namespace=namespace)
        self.data = data
        self.secret_type = secret_type
    
    def to_yaml(self) -> Dict[str, Any]:
        import base64
        encoded_data = {k: base64.b64encode(v.encode()).decode() 
                       for k, v in self.data.items()}
        
        return {
            "apiVersion": "v1",
            "kind": "Secret",
            "metadata": {
                "name": self.name,
                "namespace": self.namespace,
                "labels": self.labels
            },
            "type": self.secret_type,
            "data": encoded_data
        }


class HorizontalPodAutoscaler(Resource):
    """Kubernetes HPA for auto-scaling"""
    
    def __init__(self, name: str, target: Union[Deployment, str], 
                 min_replicas: int = 1, max_replicas: int = 10,
                 cpu_percent: int = 80, namespace: str = "default", **kwargs):
        super().__init__(kind="HorizontalPodAutoscaler", name=name, namespace=namespace)
        self.target = target.name if isinstance(target, Deployment) else target
        self.min_replicas = min_replicas
        self.max_replicas = max_replicas
        self.cpu_percent = cpu_percent
    
    def to_yaml(self) -> Dict[str, Any]:
        return {
            "apiVersion": "autoscaling/v2",
            "kind": "HorizontalPodAutoscaler",
            "metadata": {
                "name": self.name,
                "namespace": self.namespace,
                "labels": self.labels
            },
            "spec": {
                "scaleTargetRef": {
                    "apiVersion": "apps/v1",
                    "kind": "Deployment",
                    "name": self.target
                },
                "minReplicas": self.min_replicas,
                "maxReplicas": self.max_replicas,
                "metrics": [{
                    "type": "Resource",
                    "resource": {
                        "name": "cpu",
                        "target": {
                            "type": "Utilization",
                            "averageUtilization": self.cpu_percent
                        }
                    }
                }]
            }
        }


# Utility functions for common patterns
def serviceFor(deployment: Deployment, port: Optional[int] = None, 
               service_type: str = "ClusterIP") -> Service:
    """Create a service for a deployment"""
    return Service(
        name=f"{deployment.name}-service",
        target=deployment,
        port=port or deployment.port or 80,
        service_type=service_type,
        namespace=deployment.namespace
    )


def autoscaleFor(deployment: Deployment, min_replicas: int = 1, 
                max_replicas: int = 10, cpu_percent: int = 80) -> HorizontalPodAutoscaler:
    """Create autoscaling for a deployment"""
    return HorizontalPodAutoscaler(
        name=f"{deployment.name}-hpa",
        target=deployment,
        min_replicas=min_replicas,
        max_replicas=max_replicas,
        cpu_percent=cpu_percent,
        namespace=deployment.namespace
    )


def deploy(*resources: Resource, dry_run: bool = False, apply: bool = True) -> bool:
    """Deploy resources to Kubernetes cluster"""
    print(f"🚀 Deploying {len(resources)} resources...")
    
    all_yaml = []
    for resource in resources:
        yaml_content = resource.to_yaml()
        all_yaml.append(yaml_content)
        print(f"   📦 {resource.kind}: {resource.name}")
    
    if dry_run:
        print("💡 Dry run - YAML would be:")
        import yaml
        for yaml_content in all_yaml:
            print("---")
            print(yaml.dump(yaml_content, default_flow_style=False))
        return True
    
    if not apply:
        return True
    
    # Write to temporary file and apply with kubectl
    try:
        import yaml
        with tempfile.NamedTemporaryFile(mode='w', suffix='.yaml', delete=False) as f:
            for i, yaml_content in enumerate(all_yaml):
                if i > 0:
                    f.write("---\n")
                yaml.dump(yaml_content, f, default_flow_style=False)
            f.flush()
            
            # Apply with kubectl
            result = subprocess.run(['kubectl', 'apply', '-f', f.name], 
                                  capture_output=True, text=True)
            
            Path(f.name).unlink()  # Clean up temp file
            
            if result.returncode == 0:
                print("✅ Deployment successful!")
                if result.stdout:
                    print(result.stdout)
                return True
            else:
                print(f"❌ Deployment failed: {result.stderr}")
                return False
                
    except ImportError:
        print("❌ PyYAML required: pip install pyyaml")
        return False
    except FileNotFoundError:
        print("❌ kubectl not found. Please install kubectl and configure cluster access.")
        return False
    except Exception as e:
        print(f"❌ Deployment error: {e}")
        return False


# Smart deployment patterns
def webApp(name: str, image: str, replicas: int = 3, port: int = 80,
           env: Optional[Dict[str, str]] = None, auto_scale: bool = True) -> List[Resource]:
    """Deploy a complete web application with service and optional auto-scaling"""
    app = Deployment(name, image, replicas=replicas, port=port, env=env)
    service = serviceFor(app, service_type="LoadBalancer")
    
    resources = [app, service]
    
    if auto_scale:
        hpa = autoscaleFor(app, min_replicas=replicas, max_replicas=replicas * 3)
        resources.append(hpa)
    
    return resources


def database(name: str, image: str, storage: str = "10Gi", 
             env: Optional[Dict[str, str]] = None) -> List[Resource]:
    """Deploy a database with persistent storage"""
    db = StatefulSet(name, image, storage=storage, env=env)
    service = serviceFor(db)
    return [db, service]


# Example usage - Copy and modify for your needs
if __name__ == "__main__":
    print("🎯 Mini-Kuber Python Template Example")
    
    # Example 1: Simple web application
    web_resources = webApp(
        name="my-web-app",
        image="nginx:latest",
        replicas=2,
        port=80,
        auto_scale=True
    )
    
    # Example 2: Database
    db_resources = database(
        name="postgres-db",
        image="postgres:13",
        storage="20Gi",
        env={
            "POSTGRES_DB": "myapp",
            "POSTGRES_USER": "admin",
            "POSTGRES_PASSWORD": "secret123"
        }
    )
    
    # Example 3: Configuration and secrets
    config = ConfigMap("app-config", {
        "app.properties": "debug=true\nlog_level=info"
    })
    
    secret = Secret("app-secrets", {
        "api_key": "super-secret-key",
        "db_password": "ultra-secure-password"
    })
    
    # Deploy everything
    all_resources = web_resources + db_resources + [config, secret]
    
    # Uncomment to deploy (requires kubectl and cluster access)
    # deploy(*all_resources)
    
    # For testing, show what would be deployed
    deploy(*all_resources, dry_run=True)

File: mini-kuber.js
#!/usr/bin/env node

/**
 * Mini-Kuber JavaScript Template
 * Replace YAML hell with intelligent JavaScript code
 * 
 * Usage:
 * 1. Copy this file to your project
 * 2. Modify the examples at the bottom
 * 3. Run: node your-deployment.js
 * 4. Your Kubernetes resources are deployed
 * 
 * No framework to install, no complexity, just copy and adapt.
 */

const fs = require('fs');
const { execSync } = require('child_process');
const path = require('path');

class Resource {
    constructor(kind, name, namespace = 'default') {
        this.kind = kind;
        this.name = name;
        this.namespace = namespace;
        this.labels = { app: name, 'managed-by': 'mini-kuber' };
        this.annotations = {};
    }
}

class Deployment extends Resource {
    /**
     * Kubernetes Deployment made simple
     */
    constructor(name, options = {}) {
        super('Deployment', name, options.namespace || 'default');
        
        this.image = options.image || 'nginx:latest';
        this.replicas = options.replicas || 1;
        this.port = options.port;
        this.env = options.env || {};
        this.cpuRequest = options.cpuRequest || '100m';
        this.memoryRequest = options.memoryRequest || '128Mi';
        this.cpuLimit = options.cpuLimit || '500m';
        this.memoryLimit = options.memoryLimit || '512Mi';
        
        if (options.autoHeal !== false) {
            this.labels['auto-heal'] = 'true';
        }
    }
    
    toYAML() {
        const container = {
            name: this.name,
            image: this.image,
            resources: {
                requests: { cpu: this.cpuRequest, memory: this.memoryRequest },
                limits: { cpu: this.cpuLimit, memory: this.memoryLimit }
            }
        };
        
        if (this.port) {
            container.ports = [{ containerPort: this.port }];
        }
        
        if (Object.keys(this.env).length > 0) {
            container.env = Object.entries(this.env).map(([name, value]) => ({
                name, value: String(value)
            }));
        }
        
        return {
            apiVersion: 'apps/v1',
            kind: 'Deployment',
            metadata: {
                name: this.name,
                namespace: this.namespace,
                labels: this.labels,
                annotations: this.annotations
            },
            spec: {
                replicas: this.replicas,
                selector: { matchLabels: { app: this.name } },
                template: {
                    metadata: { labels: { app: this.name, ...this.labels } },
                    spec: { containers: [container] }
                }
            }
        };
    }
}

class Service extends Resource {
    /**
     * Kubernetes Service made simple
     */
    constructor(name, target, options = {}) {
        super('Service', name, options.namespace || 'default');
        
        this.target = typeof target === 'string' ? target : target.name;
        this.port = options.port || 80;
        this.targetPort = options.targetPort || (target.port || this.port);
        this.type = options.type || 'ClusterIP';
    }
    
    toYAML() {
        return {
            apiVersion: 'v1',
            kind: 'Service',
            metadata: {
                name: this.name,
                namespace: this.namespace,
                labels: this.labels,
                annotations: this.annotations
            },
            spec: {
                selector: { app: this.target },
                ports: [{
                    port: this.port,
                    targetPort: this.targetPort,
                    protocol: 'TCP'
                }],
                type: this.type
            }
        };
    }
}

class StatefulSet extends Resource {
    /**
     * Kubernetes StatefulSet for databases and stateful apps
     */
    constructor(name, options = {}) {
        super('StatefulSet', name, options.namespace || 'default');
        
        this.image = options.image;
        this.replicas = options.replicas || 1;
        this.port = options.port;
        this.storage = options.storage || '10Gi';
        this.storageClass = options.storageClass || 'standard';
        this.env = options.env || {};
    }
    
    toYAML() {
        const container = {
            name: this.name,
            image: this.image,
            volumeMounts: [{
                name: `${this.name}-storage`,
                mountPath: '/var/lib/data'
            }]
        };
        
        if (this.port) {
            container.ports = [{ containerPort: this.port }];
        }
        
        if (Object.keys(this.env).length > 0) {
            container.env = Object.entries(this.env).map(([name, value]) => ({
                name, value: String(value)
            }));
        }
        
        return {
            apiVersion: 'apps/v1',
            kind: 'StatefulSet',
            metadata: {
                name: this.name,
                namespace: this.namespace,
                labels: this.labels
            },
            spec: {
                serviceName: this.name,
                replicas: this.replicas,
                selector: { matchLabels: { app: this.name } },
                template: {
                    metadata: { labels: { app: this.name } },
                    spec: { containers: [container] }
                },
                volumeClaimTemplates: [{
                    metadata: { name: `${this.name}-storage` },
                    spec: {
                        accessModes: ['ReadWriteOnce'],
                        storageClassName: this.storageClass,
                        resources: { requests: { storage: this.storage } }
                    }
                }]
            }
        };
    }
}

class ConfigMap extends Resource {
    /**
     * Kubernetes ConfigMap for configuration
     */
    constructor(name, data, options = {}) {
        super('ConfigMap', name, options.namespace || 'default');
        this.data = data;
    }
    
    toYAML() {
        return {
            apiVersion: 'v1',
            kind: 'ConfigMap',
            metadata: {
                name: this.name,
                namespace: this.namespace,
                labels: this.labels
            },
            data: this.data
        };
    }
}

class Secret extends Resource {
    /**
     * Kubernetes Secret for sensitive data
     */
    constructor(name, data, options = {}) {
        super('Secret', name, options.namespace || 'default');
        this.data = data;
        this.type = options.type || 'Opaque';
    }
    
    toYAML() {
        const encodedData = {};
        for (const [key, value] of Object.entries(this.data)) {
            encodedData[key] = Buffer.from(String(value)).toString('base64');
        }
        
        return {
            apiVersion: 'v1',
            kind: 'Secret',
            metadata: {
                name: this.name,
                namespace: this.namespace,
                labels: this.labels
            },
            type: this.type,
            data: encodedData
        };
    }
}

class HorizontalPodAutoscaler extends Resource {
    /**
     * Kubernetes HPA for auto-scaling
     */
    constructor(name, target, options = {}) {
        super('HorizontalPodAutoscaler', name, options.namespace || 'default');
        
        this.target = typeof target === 'string' ? target : target.name;
        this.minReplicas = options.minReplicas || 1;
        this.maxReplicas = options.maxReplicas || 10;
        this.cpuPercent = options.cpuPercent || 80;
    }
    
    toYAML() {
        return {
            apiVersion: 'autoscaling/v2',
            kind: 'HorizontalPodAutoscaler',
            metadata: {
                name: this.name,
                namespace: this.namespace,
                labels: this.labels
            },
            spec: {
                scaleTargetRef: {
                    apiVersion: 'apps/v1',
                    kind: 'Deployment',
                    name: this.target
                },
                minReplicas: this.minReplicas,
                maxReplicas: this.maxReplicas,
                metrics: [{
                    type: 'Resource',
                    resource: {
                        name: 'cpu',
                        target: {
                            type: 'Utilization',
                            averageUtilization: this.cpuPercent
                        }
                    }
                }]
            }
        };
    }
}

// Utility functions for common patterns
function serviceFor(deployment, options = {}) {
    return new Service(
        `${deployment.name}-service`,
        deployment,
        {
            port: options.port || deployment.port || 80,
            type: options.type || 'ClusterIP',
            namespace: deployment.namespace,
            ...options
        }
    );
}

function deploy(...resources) {
    const options = resources[resources.length - 1];
    const actualResources = typeof options === 'object' && options.dryRun !== undefined 
        ? resources.slice(0, -1) 
        : resources;
    const dryRun = options && options.dryRun;
    const apply = options && options.apply !== false;

    console.log(`🚀 Deploying ${actualResources.length} resources...`);
    
    const allYAML = [];
    for (const resource of actualResources) {
        const yamlContent = resource.toYAML();
        allYAML.push(yamlContent);
        console.log(`   📦 ${resource.kind}: ${resource.name}`);
    }
    
    if (dryRun) {
        console.log('💡 Dry run - YAML would be:');
        const yaml = require('js-yaml');
        for (const yamlContent of allYAML) {
            console.log('---');
            console.log(yaml.dump(yamlContent));
        }
        return true;
    }
    
    if (!apply) {
        return true;
    }
    
    try {
        const yaml = require('js-yaml');
        const tempFile = path.join(require('os').tmpdir(), `mini-kuber-${Date.now()}.yaml`);
        
        let yamlString = '';
        for (let i = 0; i < allYAML.length; i++) {
            if (i > 0) yamlString += '---\n';
            yamlString += yaml.dump(allYAML[i]);
        }
        
        fs.writeFileSync(tempFile, yamlString);
        
        try {
            execSync(`kubectl apply -f ${tempFile}`, { stdio: 'inherit' });
            console.log('✅ Deployment successful!');
            return true;
        } finally {
            fs.unlinkSync(tempFile);
        }
        
    } catch (error) {
        if (error.message.includes('js-yaml')) {
            console.log('❌ js-yaml required: npm install js-yaml');
        } else if (error.message.includes('kubectl')) {
            console.log('❌ kubectl not found. Please install kubectl and configure cluster access.');
        } else {
            console.log(`❌ Deployment error: ${error.message}`);
        }
        return false;
    }
}

// Smart deployment patterns
function webApp(name, image, options = {}) {
    const replicas = options.replicas || 3;
    const port = options.port || 80;
    const autoScale = options.autoScale !== false;
    
    const app = new Deployment(name, { 
        image, 
        replicas, 
        port, 
        env: options.env,
        namespace: options.namespace 
    });
    
    const service = serviceFor(app, { type: 'LoadBalancer' });
    const resources = [app, service];
    
    if (autoScale) {
        const hpa = autoscaleFor(app, {
            minReplicas: replicas,
            maxReplicas: replicas * 3
        });
        resources.push(hpa);
    }
    
    return resources;
}

function database(name, image, options = {}) {
    const db = new StatefulSet(name, {
        image,
        storage: options.storage || '10Gi',
        env: options.env,
        port: options.port,
        namespace: options.namespace
    });
    
    const service = serviceFor(db);
    return [db, service];
}

// Factory functions for cleaner syntax
function deployment(name, options) {
    return new Deployment(name, options);
}

function service(name, target, options) {
    return new Service(name, target, options);
}

function statefulSet(name, options) {
    return new StatefulSet(name, options);
}

function configMap(name, data, options) {
    return new ConfigMap(name, data, options);
}

function secret(name, data, options) {
    return new Secret(name, data, options);
}

function hpa(name, target, options) {
    return new HorizontalPodAutoscaler(name, target, options);
}

// Export for module usage
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        Deployment,
        Service,
        StatefulSet,
        ConfigMap,
        Secret,
        HorizontalPodAutoscaler,
        serviceFor,
        autoscaleFor,
        deploy,
        webApp,
        database,
        deployment,
        service,
        statefulSet,
        configMap,
        secret,
        hpa
    };
}

// Example usage - Copy and modify for your needs
if (require.main === module) {
    console.log('🎯 Mini-Kuber JavaScript Template Example');
    
    // Example 1: Simple web application
    const webResources = webApp('my-web-app', 'nginx:latest', {
        replicas: 2,
        port: 80,
        autoScale: true
    });
    
    // Example 2: Database
    const dbResources = database('postgres-db', 'postgres:13', {
        storage: '20Gi',
        env: {
            POSTGRES_DB: 'myapp',
            POSTGRES_USER: 'admin',
            POSTGRES_PASSWORD: 'secret123'
        }
    });
    
    // Example 3: Configuration and secrets
    const config = configMap('app-config', {
        'app.properties': 'debug=true\nlog_level=info'
    });
    
    const appSecret = secret('app-secrets', {
        api_key: 'super-secret-key',
        db_password: 'ultra-secure-password'
    });
    
    // Deploy everything
    const allResources = [...webResources, ...dbResources, config, appSecret];
    
    // Uncomment to deploy (requires kubectl and cluster access)
    // deploy(...allResources);
    
    // For testing, show what would be deployed
    deploy(...allResources, { dryRun: true });
}
            ...options
        }
    );
}

function autoscaleFor(deployment, options = {}) {
    return new HorizontalPodAutoscaler(
        `${deployment.name}-hpa`,
        deployment,
        {
            minReplicas: options.minReplicas || 1,
            maxReplicas: options.maxReplicas || 10,
            cpuPercent: options.cpuPercent || 80,
            namespace: deployment.namespace,